cluster_generator/data_structures.py

```diff
@@ -1,100 +1,560 @@
-
-from cluster_generator.utils import ensure_list, ensure_ytarray, mylog
-
-
-def build_yt_dataset_fields(grid, models, domain_dimensions, centers, velocities):
-    """Builds the datastructures for initializing yt datasets"""
-    from cluster_generator.model import ClusterModel
-
-    # -- Segmenting the fields by type -- #
-    _added_fields = [
-        "density",
-        "pressure",
-        "dark_matter_density",
-        "stellar_density",
-        "gravitational_potential",
-    ]
-    _mass_weighted_fields = ["temperature"]
-    _mass_weighted_nonfields = ["velocity_x", "velocity_y", "velocity_z"]
-    units = {
-        "density": "Msun/kpc**3",
-        "pressure": "Msun/kpc/Myr**2",
-        "dark_matter_density": "Msun/kpc**3",
-        "stellar_density": "Msun/kpc**3",
-        "temperature": "K",
-        "gravitational_potential": "kpc**2/Myr**2",
-        "velocity_x": "kpc/Myr",
-        "velocity_y": "kpc/Myr",
-        "velocity_z": "kpc/Myr",
-        "magnetic_field_strength": "G",
-    }
-
-    # -- Sanity Checks -- #
-    models = ensure_list(models)
-
-    for mid, model in enumerate(models):
-        if isinstance(model, str):
-            models[mid] = ClusterModel.from_h5_file(model)
-
-    centers = ensure_ytarray(centers, "kpc")
-    velocities = ensure_ytarray(velocities, "kpc/Myr")
-
-    mylog.info("Building yt dataset structure...")
-
-    x, y, z = grid
-    fields = _added_fields + _mass_weighted_fields
-    data = {}
-    for i, p in enumerate(models):
-        xx = x - centers.d[i][0]
-        yy = y - centers.d[i][1]
-        zz = z - centers.d[i][2]
-        rr = np.sqrt(xx * xx + yy * yy + zz * zz)
-        fd = InterpolatedUnivariateSpline(p["radius"].d, p["density"].d)
-
-        # -- Managing constituent fields -- #
-        for field in fields:
-            if field not in p:  # We don't have this data, don't do anything.
-                continue
+from scipy.interpolate import dfitpack
+from tqdm import tqdm
+
+from cluster_generator import ClusterModel
+from cluster_generator.opt.structures import (
+    construct_chunks,
+    dump_field_to_hdf5,
+    renormalize,
+    unnormalize,
+)
+from cluster_generator.utils import cgparams, ensure_ytarray, mylog
+
+# from cluster_generator.opt.structures import interpolate_from_tables
+#: Field names corresponding to cases where fields may be added without concern for a weight function.
+additive_fields = [
+    "density",
+    "pressure",
+    "dark_matter_density",
+    "stellar_density",
+    "gravitational_potential",
+]
+#: Fields which need to be weighted by gas density when models are being combined.
+dens_weighted_fields = ["temperature"]  # Density weighted fields.
+#: density weighted non-fields
+dens_weighted_nonfields = {
+    "velocity_x": ("density", 0, "velocity"),
+    "velocity_y": ("density", 1, "velocity"),
+    "velocity_z": ("density", 2, "velocity"),
+    "stellar_velocity_x": ("stellar_density", 0, "velocity"),
```

Comment:
Should star and DM velocity fields be here? I actually don't think so. They're not going to be grid-based quantities. It may make sense to deposit the star and DM density fields, but not these.